### PR TITLE
Refactor 'MetadataReader'

### LINKED ISSUE #71: Refactor `MetadataReader`

### PR DESCRIPTION
**Pull request by chiaradiamarcelo**
Closes #71.

### CODE REVIEW

File: src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java [hunk 1]
```diff
@@ -91,22 +92,39 @@ public IdentifierConverter getIdentifierConverter() {
 
     @Override
     public SchemaMetadata readRemoteSchemaMetadata() {
+        return readRemoteSchemaMetadataForSelectedTables(List.of());
```
Comment (redcatbear) [inline src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java:95]: Using an empty list to say "all" is counter-intuitive.

File: src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java [hunk 2]
```diff
@@ -86,23 +76,35 @@ protected void logDropVirtualSchemaRequestReceived(final DropVirtualSchemaReques
...
     }
 
+    private SchemaMetadata getRemoteMetadata(final ExaMetadata metadata, final RefreshRequest request)
+            throws PropertyValidationException, SQLException {
+        final AdapterProperties properties = getPropertiesFromRequest(request);
+        if (request.refreshesOnlySelectedTables()) {
+            final List<String> tables = request.getTables();
+            return readMetadata(properties, tables, metadata);
+        }
+        return readMetadata(properties, metadata);
```
Comment (redcatbear) [inline src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java:96]:
```suggestion
        }
        else {
            return readMetadata(properties, metadata);
        }
```
Mind the symmetry. Also this conveys better that you thought about the "else" case.

File: src/test/java/com/exasol/adapter/jdbc/BaseTableMetadataReaderTest.java [hunk 3]
```diff
@@ -89,7 +86,7 @@ void testWarnIfTableScanResultsAreEmpty() throws SQLException {
         mockTableCount(this.tablesMock, 0);
         final CapturingLogHandler capturingLogHandler = new CapturingLogHandler();
         Logger.getLogger("com.exasol").addHandler(capturingLogHandler);
-        createDefaultTableMetadataReader().mapTables(this.tablesMock, Optional.empty());
+        createDefaultTableMetadataReader().mapTables(this.tablesMock, List.of());
```
Comment (redcatbear) [inline src/test/java/com/exasol/adapter/jdbc/BaseTableMetadataReaderTest.java:89]:
```suggestion
        createDefaultTableMetadataReader().mapTables(this.tablesMock, Collections.emptyList());
```

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java (hunk 1 above)
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java:95]:
```suggestion
        return readRemoteSchemaMetadataForSelectedTables(Collections.emptyList());
```

Comment (chiaradiamarcelo) [reply src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java]: Agreed to do it in a next PR as we need to change another interface

File: src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java [hunk 4]
```diff
@@ -91,22 +91,39 @@ public IdentifierConverter getIdentifierConverter() {
...
-            final String adapterNotes = SchemaAdapterNotesJsonConverter.getInstance()
-                    .convertToJson(getSchemaAdapterNotes());
-            final DatabaseMetaData remoteMetadata = this.connection.getMetaData();
-            final List<TableMetadata> tables = extractTableMetadata(remoteMetadata, Optional.empty());
-            return new SchemaMetadata(adapterNotes, tables);
+            return this.getSchemaMetadata(selectedTables);
         } catch (final SQLException exception) {
-            throw new RemoteMetadataReaderException(
-                    "Unable to read remote schema metadata. SQL error: " + exception.getMessage(), exception);
+            throw new RemoteMetadataReaderException("Unable to read remote schema metadata.", exception);
```
Comment (redcatbear) [inline src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java:106]: Don't remove the cause workaround. Older versions of the UDF don't display the exception stack.

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java (hunk 4 above)
Comment (chiaradiamarcelo) [inline src/main/java/com/exasol/adapter/jdbc/AbstractRemoteMetadataReader.java:106]:
```suggestion
            throw new RemoteMetadataReaderException(
                    "Unable to read remote schema metadata. SQL error: " + exception.getMessage(), exception);
```

**Review by jakobbraun — APPROVED** [review]
looks good

**Review by redcatbear — APPROVED** [review]

**Pull request merged**